atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/monitoring/TimestampTracker.java

```diff
@@ -0,0 +1,114 @@
+    @VisibleForTesting
+    static final Duration CACHE_INTERVAL = Duration.ofSeconds(10L);
+
+    private final MetricsManager metricsManager = new MetricsManager();
+    private final Clock clock;
+
+    public TimestampTracker(Clock clock) {
+        this.clock = clock;
+    }
+
+    public static TimestampTracker createNoOpTracker() {
+        return new TimestampTracker(Clock.defaultClock());
+    }
+
+    public static TimestampTracker createWithDefaultTrackers(TimelockService timeLockService, Cleaner cleaner) {
+        TimestampTracker tracker = new TimestampTracker(Clock.defaultClock());
+        tracker.registerTimestampForTracking("timestamp.fresh",
+                TimestampTracker.ignoreNotInitializedException(timeLockService::getFreshTimestamp));
+        tracker.registerTimestampForTracking("timestamp.immutable",
+                TimestampTracker.ignoreNotInitializedException(timeLockService::getImmutableTimestamp));
+        tracker.registerTimestampForTracking("timestamp.unreadable",
+                TimestampTracker.ignoreNotInitializedException(cleaner::getUnreadableTimestamp));
+        return tracker;
+    }
+
+    @VisibleForTesting
+    void registerTimestampForTracking(String shortName, Supplier<Long> supplier) {
+        metricsManager.registerMetric(
+                TimestampTracker.class,
+                shortName,
+                createCachingTimestampGauge(shortName, supplier));
+    }
+
+    private Gauge<Long> createCachingTimestampGauge(String shortName, Supplier<Long> supplier) {
+        return new CachedGauge<Long>(clock, CACHE_INTERVAL.getSeconds(), TimeUnit.SECONDS) {
+            AtomicLong upperBound = new AtomicLong(Long.MIN_VALUE);
+
+            @Override
+            protected Long loadValue() {
+                try {
+                    // Note that this gauge is only an approximation, because some timestamps can go backwards.
+                    return upperBound.accumulateAndGet(supplier.get(), Math::max);
+                } catch (Exception e) {
+                    long timestampToReturn = upperBound.get();
+                    log.info("An exception occurred when trying to update the {} timestamp for tracking purposes."
+                                    + " Returning the last known value of {}.",
+                            SafeArg.of("timestampName", shortName),
+                            SafeArg.of("timestamp", timestampToReturn),
+                            e);
+                    return timestampToReturn;
+                }
+            }
+        };
+    }
+
+    private static Supplier<Long> ignoreNotInitializedException(Supplier<Long> metricFunction) {
+        return () -> {
+            try {
+                return metricFunction.get();
+            } catch (NotInitializedException exception) {
+                // Ignore not initialized exception, as we don't want to have metrics from when the services are not
+                // initialized.
+                return 0L;
```

Comment:
Hmm, ideally I was hoping for us to not record a value in this case (by making the timestamp tracker itself be async initialized) but I guess this works too.